feat(assets): allow `new URL` to resolve package assets

kherock opened the pull request:
<!-- Thank you for contributing! -->

### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->
This change allows modules to resolve asset URLs for assets located underneath a package path. E.g. `new URL('@scope/component/assets/logo.png', import.meta.url)`

close #6918 

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->
I Implemented this in a backwards-compatible way by using a resolver which prefers relative paths and makes the root path resolve to the public base path.

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

kherock committed:
test(assets): add fixture for `new URL("/...", import.meta.url)`

This tests existing behavior for obtaining an asset's URL from the public base path.

kherock committed:
feat(assets): allow `new URL` to resolve package assets

kherock commented:
I'm facing an issue where I can't seem to get my changes to get package paths to resolve while running the dev server. I'm having trouble attaching to the dev server worker or getting any logs to diagnose the problem. I'd appreciate any insight on how to debug this!

poyoho commented:
I'm not sure if this is necessary, because the type of new URL is 

```ts
constructor (input: string, base?: string | URL);
```

the second params is base, if you support the `new URL('@/hello.js', import.meta.url)` I think it's a little confusing

bluwy commented on packages/vite/src/node/plugins/assetImportMetaUrl.ts:
```diff
@@ -62,7 +62,10 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
 
           const url = rawUrl.slice(1, -1)
-          const file = path.resolve(path.dirname(id), url)
+          const resolved = await this.resolve(url, id)
+          const file = resolved
+            ? resolved.id
+            : path.resolve(path.dirname(id), url)
```
Second @poyoho's comment that we could evaluate `new URL('@some/package')` without `import.meta.url` too.

And with that in mind, we could handle this part of the code differently. If the user adds a `import.meta.url`, we'll use the `path.resolve(path.dirname(id), url)` method; If they don't have `import.meta.url`, we use the `this.resolve(url, id)` method.

poyoho replied:
oh, this is right. `new URL('/hello.js')` is only appear on file in `publish` dirs?

bluwy replied:
I think url that starts with `/` would be resolved from the project root like [glob imports](https://vitejs.dev/guide/features.html#glob-import)

kherock replied:
I forgot to push up changes that I had amended locally. If the URL starts with `/`, the existing behavior is to resolve a file from the public base dir. I've added a test for this scenario

poyoho commented:
> I'm facing an issue where I can't seem to get my changes to get package paths to resolve while running the dev server. I'm having trouble attaching to the dev server worker or getting any logs to diagnose the problem. I'd appreciate any insight on how to debug this!

because the plugin only run in build mode. 

https://github.com/vitejs/vite/blob/fc37bdc1fb817983b91d556b00dd479447362589/packages/vite/src/node/build.ts#L317

kherock commented:
> You can refer to workerImportMetaUrl
As far as I can tell, this plugin is for transforming code that uses `new Worker(new URL(...))`. I could add the same support to that plugin, but it doesn't cover assets.

kherock commented on packages/playground/assets/__tests__/assets.spec.ts:
```diff
@@ -234,6 +234,12 @@ test('new URL(..., import.meta.url)', async () => {
   expect(await page.textContent('.import-meta-url')).toMatch(assetMatch)
 })
 
+test('new URL(/..., import.meta.url)', async () => {
+  expect(await page.textContent('.import-meta-url-base-path')).toMatch(
+    iconMatch
+  )
+})
+
```
On `main`, fixture I added here actually doesn't work for both sets of tests.

During build, if the asset is small enough, it's resolved and optimized into a Data URI despite being in the public folder. Otherwise, it is resolved to be relative the the public base path.

During serve, the import plugin isn't run at all, and this resolves to the root of the site.

Which one of these is more correct? I think ideally, the root path should always resolve from the public base directory. If developers want it to be from the root of the site (as it does with serve mode), they should just omit `import.meta.url`.

kherock committed:
feat(worker): allow `new URL` to resolve workers from package aliases

poyoho commented on packages/playground/assets/index.html:
```diff
@@ -276,6 +284,16 @@ <h3 class="raw-query"></h3>
   text('.import-meta-url', metaUrl)
   document.querySelector('.import-meta-url-img').src = metaUrl
 
+  const metaUrlDep = new URL('@/asset.png', import.meta.url)
```
What do you think about the https://github.com/vitejs/vite/pull/7837#issuecomment-1104672574. I think is better to remove `import.meta.url` in absolute path.

Maybe you can use this regexp

```js
const assetImportMetaUrlRE =
  /\bnew\s+URL\s*\(\s*('[^']+'|"[^"]+"|`[^`]+`)\s*(,|,\s*import\.meta\.url){0,1}\s*\)/g
```

https://regex101.com/r/mSdtWD/1

kherock replied:
Keeping `import.meta.url` for resolving package assets is important since it gives Vite parity with webpack in this scenario.

When `import.meta.url` isn't included, I'm wondering if a custom protocol could be used instead. For example `new URL('vite:@/asset.png'). I kind of like this because this allows the asset to use a valid absolute URL and is similar to how Node defines its `node:` URLs for its builtins.

bluwy replied:
For now I think we can also have `import.meta.url` as is. It wouldn't be used when resolving absolute or aliased path, but perhaps we can have another discussion to allow without `import.meta.url` too.

poyoho commented on packages/playground/assets/__tests__/assets.spec.ts:
```diff
@@ -234,6 +234,16 @@ test('new URL(..., import.meta.url)', async () => {
   expect(await page.textContent('.import-meta-url')).toMatch(assetMatch)
 })
 
+test('new URL(@/..., import.meta.url)', async () => {
```
nitpick 

```diff
- test('new URL(@/..., import.meta.url)', async () => {
+ test('new URL("@/...", import.meta.url)', async () => {
```

poyoho commented on packages/vite/src/node/plugins/worker.ts:
```diff
@@ -278,10 +278,11 @@ export function webWorkerPlugin(config: ResolvedConfig): Plugin {
-    renderChunk(code) {
-      if (isWorker && code.includes('import.meta.url')) {
-        return code.replace('import.meta.url', 'self.location.href')
+    resolveImportMeta(property) {
```
good point to fix `import.meta.url` in worker, but I think this is no part of the PR.

kherock replied:
I'll revert this, let me know if you want me to make another PR for this update!

kherock replied:
I actually can't revert this and keep my new tests for the worker support. The old test suite was missing the check for using `new URL` to construct a worker and it fails without this fix.

poyoho replied:
OK， keep it. I open PR to test it 😊

bluwy replied:
This is indeed a very clean solution 👍

kherock replied:
I'm reverting [6a0384c](https://github.com/vitejs/vite/pull/7837/commits/6a0384c2bfd135224cf0e8d920f07e9f89a5beb6) since it looks like this was resolved upstream by other means?

kherock committed:
test(worker): add fixture for `new URL(..., import.meta.url)`

This test fails because the worker plugin's renderChunk hook runs its
import.meta.url transformation after Rollup already has transformed it
with importMeta.renderFinalMechanism.

poyoho commented:
fix: #7779

bluwy commented on packages/vite/src/node/plugins/assetImportMetaUrl.ts:
```diff
@@ -16,6 +17,8 @@ import { emptyString } from '../cleanString'
+
   return {
     name: 'vite:asset-import-meta-url',
     async transform(code, id, options) {
```
GitHub doesn't allow me to highlight line41:

```js
const rawUrl = code.slice(urlStart, urlEnd)
```

In development mode if `rawUrl` starts with a `.`, I think we can `continue` the while loop, as we leave the code as is for the browser to resolve like the current Vite behaviour.

kherock replied:
That makes sense. I was also wondering if it makes sense for us to turn off `preferRelative: true` in the resolver. I decided not to since I believe it's unintuitive for `new URL("./foo.png", import.meta.url)` to have different behavior `new URL("foo.png", import.meta.url)`.

bluwy replied:
Yeah we can leave `preferRelative: true` as is so it's consistent with the glob behaviour.

bluwy commented on packages/vite/src/node/plugins/assetImportMetaUrl.ts:
```diff
@@ -62,16 +65,31 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
-          // Get final asset URL. Catch error if the file does not exist,
-          // in which we can resort to the initial URL and let it resolve in runtime
-          const builtUrl = await fileToUrl(file, config, this).catch(() => {
+          assetResolver ??= config.createResolver({
+            root: config.publicDir,
+            extensions: [],
+            mainFields: [],
+            tryIndex: false,
+            preferRelative: true
+          })
```
I think we should match the resolver used for `import.meta.glob('@/alias')` too?

https://github.com/vitejs/vite/blob/0d2089c6876bcc04a46c3f7947ce4effc91dd9fd/packages/vite/src/node/plugins/importAnalysis.ts#L111-L115

Not sure if that makes sense.

Re `root: config.publicDir`, if we follow above, this would be removed. If the user wants to use a public file, shouldn't they specify the string directly without `new URL`?

kherock replied:
For public files—they could use a relative path, though if they wanted to resolve it to an absolute URL, they'd use `new URL('path/to/file', self.location.href)` or `new URL('/baseDir/path/to/file', self.location.href)`. However, a `<base href>` tag can interfere with the former, and the latter requires that you embed the output base directory every time.

Regardless I think the existing `root` I have should probably be removed because it prevents other directories that might be at the project root from resolving—for example, the project root might have `img src static`, with code that uses `new URL('/img/logo.png', import.meta.url)`. The resolver would attempt to find the file at `/static/img/logo.png` and fail.

Off-topic, but why does `import.meta.glob` have `preferRelative: true`? This seems at odds with the Vite docs
> The glob patterns are treated like import specifiers: they must be either relative (start with `./`) or absolute (start with `/`, resolved relative to project root) or an alias path

bluwy replied:
I believe `preferRelative: true` is so that `new URL("./foo.png", import.meta.url)` is the same as `new URL("foo.png", import.meta.url)` like you mentioned above. It's a very rare case and could be safe to remove too, but I think it's harmless to be there too.

I don't see how it's at odds with the docs though, but here's how `preferRelative` changes things in the resolver:

https://github.com/vitejs/vite/blob/3edd1af56e980aef56641a5a51cf2932bb580d41/packages/vite/src/node/plugins/resolve.ts#L170

It adds another check to see if it can be resolved relatively (though we would already have a fast-path for it).

kherock replied:
I pointed out that the existing glob behavior is at odds with the docs because right now, with `preferRelative: true`, relative imports do not have to start with `./`.

I just wanted to bring it to someone's attention in case it mattered. `preferRelative: true` definitely makes sense for these asset URLs here, though.

bluwy replied:
I see, yeah that's true. We can look into this in the future to make it more restrictive. I think the implementation for `transformImportGlob` could be improved too to use a single resolver.

bluwy commented on packages/vite/src/node/plugins/assetImportMetaUrl.ts:
```diff
@@ -62,16 +65,31 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+            tryIndex: false,
+            preferRelative: true
+          })
+          const resolved = await assetResolver(url, id)
+          if (!resolved) {
```
I think we can optimize this like above, when the url starts with `.`, we can result to the simpler `const file = path.resolve(path.dirname(id), url)` like before.

bluwy commented on packages/vite/src/node/plugins/assetImportMetaUrl.ts:
```diff
@@ -62,16 +65,31 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+          }
+          // Get final asset URL. If the file does not exist,
+          // we fall back to the initial URL and let it resolve in runtime
+          const builtUrl = resolved
+            ? await fileToUrl(
+                isParentDirectory(normalizePath(config.publicDir), resolved)
+                  ? url
+                  : resolved,
+                config,
+                this
+              )
+            : url
```
I think using an `if` block would make this easier to read. Also have a question with the `config.publicDir` mentioned above.

bluwy commented on packages/vite/src/node/plugins/workerImportMetaUrl.ts:
```diff
@@ -144,18 +144,27 @@ export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
           )
-          const file = path.resolve(path.dirname(id), rawUrl.slice(1, -1))
-          let url: string
+          const url = rawUrl.slice(1, -1)
+          workerResolver ??= config.createResolver({
+            tryIndex: false,
+            preferRelative: true
+          })
+          const file = (await workerResolver(url, id)) ?? url
```
Would be great to apply the same `.` optimization here too.

kherock replied:
I can add the optimization but I just wanted to point out that I'll need to specify `extensions: []` in the resolver to keep consistent behavior in dev and build modes.

bluwy replied:
Sounds good 👍

bluwy reviewed (changes requested):
Thanks for bringing this to a great state so far. I have some suggestions and questions below. Feel free to let me know if you have any concerns!

kherock committed:
perf: optimize URL constructors with relative paths

kherock commented:
I'm a bit stuck on this until I can set up my Windows environment to debug the issue with calculating a relative path from the public root. I should have some time soon but would definitely appreciate some help!

sapphi-red commented:
Maybe this implements #6918 as well?

kherock commented:
@sapphi-red this should! I'll work on rebasing this before it gets any more stale

poyoho commented on packages/vite/src/node/plugins/workerImportMetaUrl.ts:
```diff
@@ -117,22 +113,37 @@ export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
-            url = await fileToUrl(cleanUrl(file), config, this)
-            url = injectQuery(url, WORKER_FILE_ID)
-            url = injectQuery(url, `type=${workerType}`)
+            builtUrl = await fileToUrl(file, config, this)
```
why remove `cleanUrl(file)`, I think it need to avoid the same URL query

kherock replied:
This was accidentally lost in the rebase, it's back now!

poyoho commented on packages/vite/src/node/plugins/workerImportMetaUrl.ts:
```diff
@@ -117,22 +113,37 @@ export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+          s.overwrite(
+            urlIndex,
+            urlIndex + exp.length,
+            `new URL(${JSON.stringify(builtUrl)}, self.location)`,
```
I don't know why here need to `new URL(${JSON.stringify(builtUrl)}, self.location)`, thank you very much if you could tell me

kherock replied:
Really I just added this for consistency with the asset loader. The only edge case I can think of where this might have an effect is when the page has a `<base href="https://some-external-origin.com>` tag. Including `self.location` ensures that the asset/worker is resolved relative to the current origin instead of wherever the `<base>` tag is referencing.

poyoho commented:
And please run `pnpm format` to pass the lint ci

kherock commented:
Sorry for the delay @poyoho. This is ready again!

bluwy reviewed (changes requested):
Following https://github.com/vitejs/vite/pull/7837#discussion_r854776344, I think we should still resolve absolute path like `new URL('/path/to/something', import.meta.url)` to be from the `root` instead of the `publicDir`, so it reflects the resolve logic for glob imports. Not sure if this is resolved in a conversation somewhere, but I don't think the code currently handles this.

The rest looks great though!

kherock commented:
> Following [#7837 (comment)](https://github.com/vitejs/vite/pull/7837#discussion_r854776344), I think we should still resolve absolute path like `new URL('/path/to/something', import.meta.url)` to be from the `root` instead of the `publicDir`, so it reflects the resolve logic for glob imports. Not sure if this is resolved in a conversation somewhere, but I don't think the code currently handles this.
> 
> The rest looks great though!

@bluwy I do believe this is the case already. The resolver I created uses the project root, and then has a special fallback `??=` to allow `/...` routes to resolve from files from the public dir.

https://github.com/vitejs/vite/blob/899caaf278ca8a030ba59c17cc6bf89c8ebc2c70/packages/vite/src/node/plugins/assetImportMetaUrl.ts#L79-L88

In this way, both `new URL('/static/icon.png', import.meta.url)` and `new URL('/icon.png', import.meta.url)` work when `/static` is the public base dir. The former should ideally generate a warning, but I don't want to copy-paste this large chunk of code into two plugins - I'd appreciate if someone else could do that refactoring 😅

https://github.com/vitejs/vite/blob/f6bd317ebe2f6828429f520b83d14a139ef0208c/packages/vite/src/node/server/middlewares/transform.ts#L118-L147

kherock commented:
@bluwy anything I can do here? I should be able to find time to rebase this soon.

kherock commented:
This is ready again, would really like to see this released in the near future!

bluwy reviewed (approved):
Sorry for the lack of response @kherock. My brain wasn't up for recalling all the context of this PR until now 😦  Thanks for the reply previously, that make sense to me!

This LGTM. I think we can bring this in 3.1

kherock commented:
Are the macOS tests flaky in CI? Neither of the last two runs seem to be related to these changes

bluwy commented:
Looks like forth times a charm. I hope this isn't caused by this PR, plus the test failure logs seem unrelated.

kherock commented:
@poyoho @bluwy ETA on having this merged? I'd like to avoid spending more time rebasing this one if it can be helped 😅

bluwy commented:
@kherock I think this is a great feature, but we'd need a team meeting to confirm it. It's planned for 3.2 now, but if there's any merge conflicts in the future I can help with resolving it.
